From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Rowan Goemans <rowan@example.com>
Date: Wed, 15 Apr 2026 00:00:00 +0200
Subject: [PATCH] Fix side-effect loss when slicing AstExprStmt array
 expressions

diff --git a/src/V3Slice.cpp b/src/V3Slice.cpp
index 81e622ee2..092f7ba24 100644
--- a/src/V3Slice.cpp
+++ b/src/V3Slice.cpp
@@ -283,16 +283,6 @@ class SliceVisitor final : public VNVisitor {
             AstNodeAssign* const newp
                 = nodep->cloneType(cloneAndSel(nodep->lhsp(), elements, elemIdx, elemIdx != 0),
                                    cloneAndSel(nodep->rhsp(), elements, elemIdx, elemIdx != 0));
-            if (elemIdx == 0) {
-                nodep->foreach([this](AstExprStmt* const exprp) {
-                    // Result expression is always evaluated to the same value, so the statements
-                    // can be removed once they were included in the expression created for the 1st
-                    // element.
-                    AstNodeExpr* const resultp = exprp->resultp()->unlinkFrBack();
-                    exprp->replaceWith(resultp);
-                    VL_DO_DANGLING(pushDeletep(exprp), exprp);
-                });
-            }
             UINFOTREE(9, newp, "", "new");
             newlistp = AstNode::addNext(newlistp, newp);
         }
@@ -356,17 +346,6 @@ class SliceVisitor final : public VNVisitor {
                 T_NodeBiop* const clonep = new T_NodeBiop{
                     nodep->fileline(), cloneAndSel(nodep->lhsp(), elements, elemIdx, elemIdx != 0),
                     cloneAndSel(nodep->rhsp(), elements, elemIdx, elemIdx != 0)};
-                if (elemIdx == 0) {
-                    nodep->foreach([this](AstExprStmt* const exprp) {
-                        // Result expression is always evaluated to the same value, so the
-                        // statements can be removed once they were included in the expression
-                        // created for the 1st element.
-                        AstNodeExpr* const resultp = exprp->resultp()->unlinkFrBack();
-                        exprp->replaceWith(resultp);
-                        VL_DO_DANGLING(pushDeletep(exprp), exprp);
-                    });
-                }
-
                 if (!logp) {
                     logp = clonep;
                 } else {
-- 
2.49.0

